biz.aQute.repository/src/aQute/p2/provider/P2Impl.java

```diff
@@ -50,21 +51,40 @@ public P2Impl(HttpClient c, URI base, PromiseFactory promiseFactory) throws Exce
 
 	private URI normalize(URI base) throws Exception {
 		String path = base.getPath();
-		if (path.endsWith("/"))
+		if (path == null || path.endsWith("/"))
 			return base;
 
 		return new URI(base.toString() + "/");
 	}
 
-	public List<Artifact> getArtifacts() throws Exception {
+	@Override
+	public List<Artifact> getAllArtifacts() throws Exception {
 		Set<URI> cycles = Collections.newSetFromMap(new ConcurrentHashMap<URI, Boolean>());
-		return getArtifacts(cycles, base).getValue();
+		List<Artifact> value = getArtifacts(cycles, base).getValue();
+		return value;
+	}
+
+	// For backward compatibility reasons
+	// this method is now called getBundles(), it ignores non osgi.bundle
+	// artifacts
+
+	public List<Artifact> getArtifacts() throws Exception {
+		return getBundles();
 	}
 
 	private Promise<List<Artifact>> getArtifacts(Set<URI> cycles, URI uri) {
 		if (!cycles.add(uri)) {
-			return promiseFactory
-				.failed(new IllegalStateException("There is a cycle in the p2 setup : " + cycles + " -> " + uri));
+			return promiseFactory.resolved(Collections.emptyList());
+
+			//
+			// TODO discuss with BJ
```

Comment:
When will this discussion occur?

Reply:
We can do it here. I think the cycle detection should not throw an exception. Aborting that branch of the tree suffices as far as I can see. Do I miss something?